proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,562 @@
+        -- ...
+        TupleArgKind# [r1, ..., r64] = Tuple64 (TYPE r1) ... (TYPE r64)
+        TupleArgKind# other = TypeError (ShowType (Length other) :<>: Text " is too large; the maximum size of a tuple is 64.")
+
+      type Tuple# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (TupleRep reps)
+      type family Tuple# ts where
+        Tuple# () = Unit#
+        Tuple# a = Solo# a     -- see below
+        Tuple# (a, b) = Tuple2# a b
+        Tuple# (a, b, c) = Tuple3# a b c
+        -- ...
+        Tuple# (a, b, ..., bk, bl) = Tuple64# a b ... bk bl
+        Tuple# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a tuple is 64.")
+
+      type Sum# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (TupleRep reps)
+      type family Sum# ts where
+        Sum# () = TypeError (Text "GHC does not support empty unboxed sums. Consider Data.Void.Void instead.")
+        Sum# a = TypeError (Text "GHC does not support unary unboxed sums. Consider Data.Tuple.Solo# instead.")
+        Sum# (a, b) = Sum2# a b
+        Sum# (a, b, c) = Sum3# a b c
+        -- ...
+        Sum# (a, b, ..., bk, bl) = Sum64# a b ... bk bl
+        Sum# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a sum is 64.")
+
+  This creates a uniform mixfix syntax for boxed tuples, constraint tuples, unboxed tuples, and unboxed sums.
+  Handling singletons is a bit interesting:
+
+  ``TupleArgKind t_or_c 1`` is defined to be ``t_or_c``, *not* ``Solo t_or_c``, as you would
+  otherwise expect. This is because we expect to see e.g. ``Tuple (Int)``, not ``Tuple (MkSolo Int)``.
+  (Actually, we would probably not expect these at all, but the current design allows us to be forgiving
+  during refactoring.)
+
+  In ``Tuple``, we see ``Tuple a = a``, which looks like a universally applicable equation. It is
+  not. Because ``a :: Type`` (resp. ``a :: Constraint``) we learn the invisible argument to ``Tuple``
+  (resp. ``Constraints``) must be ``1``, and thus this equation fires only when the argument to ``Tuple``
+  is not a tuple of types (resp. constraints).
+
+  A beautiful, unexpected consequence of this design is that it aids migration. Now, with or without
+  ``-XListTuplePuns``, a user can write e.g. ``Tuple (Int, Bool, Double)``, and this will be the types
+  of e.g. ``(1, True, 3.14)``. With ``-XListTuplePuns``, the argument to ``Tuple`` will be the tuple
+  type, of kind ``Type``. That is, the type will really be understood as ``Tuple (Tuple3 Int Bool Double)``.
+  The "``1``" equation fires, reducing to ``Tuple3 Int Bool Double``. On the other hand, with
+  ``-XNoListTuplePuns``, the user's type is understood as ``Tuple @3 (Int, Bool, Double)``, and the "``3``"
+  equation fires, reducing to ``Tuple3 Int Bool Double`` -- the same answer! And so, all users
+  can write ``Tuple`` before their tuples and not get hurt.
```

Comment:
This part needs to be indented by one additional space, currently it is not displayed correctly.